# ![image](docs/image.jpeg) Amaranth: Single-Cell Transcript Assembler

<a href="http://bioconda.github.io/recipes/amaranth-assembler/README.html"><img src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" alt="install with bioconda"></a> <a href="https://github.com/Shao-Group/amaranth/releases"><img src="https://img.shields.io/github/v/tag/Shao-Group/Amaranth" alt="GitHub tag (latest SemVer)"></a> <a href="https://github.com/Shao-Group/amaranth/blob/master/LICENSE"><img src="https://img.shields.io/github/license/Shao-Group/Amaranth" alt="GitHub License"></a>

Amaranth is a reference-based transcript assembler specifically optimized for single-cell RNA-seq data. The development of Amaranth has been based on the [Scallop2](https://github.com/Shao-Group/scallop2)/[Scallop](https://github.com/Kingsford-Group/scallop) assembler series.

# Installation

It is recommended to install with `conda`. 

```bash
conda install bioconda::amaranth-assembler
```

Otherwise, to download pre-compiled binary for Linux, please check out GitHub Releases (`amaranth-{version}-linux` in [Latest Release](https://github.com/Shao-Group/amaranth/releases/latest)). To install from source code, please read [INSTALL.md](./INSTALL.md).

# Usage

The usage of `amaranth` is:

```
./amaranth -i <input.bam> -o <output> [options]
```

The `input.bam` is the read alignment file generated by an RNA-seq aligner. 

To correctly assemble single-cell RNA-seq reads (e.g. Smart-seq3), the bam file should have SAM optional field tag `BC` which stores cell barcodes and tag `UB` which stores UMI barcodes. See bam data in `example` directory as an example.

If you want to assemble each single cell independently, run one command for each cell. If you want to perform meta-assembly (which leverages information across all cells to improve individual cell assemblies), see [Meta-assembly Usage](#Meta-assembly-Usage).

Make sure that the bam file is sorted; otherwise run `samtools` to sort it:

```
samtools sort input.bam > input.sort.bam
```

The reconstructed transcripts shall be written as gtf format into `output.gtf`.

# Meta-assembly Usage

The usage of `amaranth` for meta-assembly is:

```bash
./amaranth --meta -i <merged.bam> -o <output> [options]
```

The argument `--meta` must be supplied to perform meta-assembly. 

The `merged.bam` is the **sorted** and **merged** read alignment file of all cells. The bam file should have SAM optional field tag `BC` which stores cell barcodes and tag `UB` which stores UMI barcodes.

If user has separate sorted read alignment files of each single cell, use `samtools` to merge them.

```bash
# merge n bam files with 32 threads
samtools merge -@32 -o merged.bam 1.bam 2.bam ... n.bam
```

The reconstructed transcripts for each cell will be written as gtf format into `output.<BC>.gtf`, where `<BC>` is the cell barcode (sam tag `BC`) of each cell. The meta-assembly (union of transcripts from all cells) will be written into `output.meta.gtf`.

To achieve the best performance, it is recommend to union the a cell's transcripts from both amaranth's meta-assembly run and individual  (as a single cell) assembly run. Union of transcripts can be done using tools such as [TACO](https://tacorna.github.io/) or [gtfmerge](https://github.com/Shao-Group/rnaseqtools?tab=readme-ov-file#gtfmerge).

## Parameters

Here is a list of supported parameters. Please refer to additional explanations below the table.

 Parameters | Default Value | Description
 ------------------------- | ------------- | ----------
 --help  | | print usage of amaranth and exit
 --version | | print version of amaranth and exit
 --meta | not used | to perform meta-assembly. 
 --preview | | show the inferred `library_type` and exit
 --verbose | 1 | chosen from {0, 1, 2}
 -f/--transcript_fragments    | | file to which the assembled non-full-length transcripts will be written to
 --library_type               | empty | chosen from {empty, unstranded, first, second}; If empty, Amaranth will try to infer automatically. 
 --assemble_duplicates		  | 10 | the number of consensus runs of the decomposition
 --min_transcript_coverage    | 1.5 | the minimum coverage required to output a multi-exon transcript
 --min_single_exon_coverage   | 20 | the minimum coverage required to output a single-exon transcript
 --min_transcript_length_base      |150 | the minimum base length of a transcript
 --min_transcript_length_increase  | 50 | the minimum increased length of a transcript with each additional exon
 --min_mapping_quality        | 1 | ignore reads with mapping quality less than this value
 --max_num_cigar              | 1000 | ignore reads with CIGAR size larger than this value
 --min_bundle_gap             | 100 | the minimum distances required to start a new bundle
 --min_num_hits_in_bundle     | 5 | the minimum number of reads required in a bundle
 --min_flank_length           | 3 | the minimum match length required in each side for a spliced read
 --min_splice_boundary_hits    | 1 | the minimum number of spliced reads required to support a junction
 --min-umi-reads-bundle | 1 | (int) Bundle with less UMI reads than this threshold will be ignored
 --min-umi-ratio-bundle           | 0.0           | (float) Bundle with lower UMI reads ratio than this threshold will be ignored
 --both-umi-support               | not used      | If set a bundle need to satisfy both `min-umi-reads-bundle` and `min-umi-ratio-bundle`. Otherwise, satisfy either of them is ok
 --min-umi-reads-start-exon | 1 | (int) minimum number of UMI reads support of the first exon in a valid transcript
 --remove-retained-intron          | used          |
 --no-remove-retained-intron       | not used      |
 --max-ir-part-ratio-v | 0.5 | the ratio threshold of retained node to skip edge for partial introns (if greater than threshold, consider true transcript)
 --max-ir-part-ratio-e | 0.5 | the ratio threshold of retained node's edge to skip edge for partial introns (if greater than threshold, consider true transcript)
 --max-ir-full-ratio-v | 1.0 | the ratio threshold of retained node to skip edge for full introns (if greater than threshold, consider true transcript)
 --max-ir-full-ratio-e | 0.5 | the ratio threshold of retained node's edge to skip edge for full introns (if greater than threshold, consider true transcript)
 --max-ir-full-ratio-i | 10.0 | the ratio threshold of retained node to its own edge for full introns (if greater than threshold, consider true RETENTION)
 --max-ir-umi-support-full | 3 | (int) maximum number of UMI reads to support a partial exon rather than full intron retention
 --max-ir-umi-support-part | 5 | (int) maximum number of UMI reads to support a partial exon rather than partial intron retention
 --remove-pcr-duplicates          | 1             | 0 (not remove) or 1 (remove w.r.t alignment coordinates and CIGAR string)
 --min-cb-ratio | 0.3 | (float) for meta-assembly, minimum ratio of exons in a transcript supported by a cell's barcode 


1. For `--verbose`, 0: quiet; 1: one line for each splice graph; 2: details of graph decomposition.
3. `--min_transcript_coverage` is used to filter lowly expressed transcripts: amaranth will filter
out transcripts whose (predicted) raw counts (number of moleculars) is less than this number.
4. `--min_transcript_length_base` and `--min_transcript_length_increase` is combined to filter
short transcripts: the minimum length of a transcript is given by `--min_transcript_length_base`
\+ `--min_transcript_length_increase` * num-of-exons-in-this-transcript. Transcripts that are less
than this number will be filtered out.



# Example

Example test data is provided in `example/`. Users can use the following command to do a basic example run.

```bash
amaranth -i example/example-input.bam -o test_output
```

The assembled transcripts will be in `test_output.gtf`.
